Documentation/design/renderphase.md

```diff
@@ -0,0 +1,129 @@
+# Render
+
+## Goals
+
+1. Render is not required to be run separately. This is exposed as a means of allowing users a deeper entry point for install-time customization.
+
+2. This renders all the assets necessary to bring up a cluster on to disk. This allows users to customize assets without over-parameterizing the InstallConfig.
+
+3. Allow user to commit the assets that describe a cluster.
+
+## Non-goals
+
+1. Render doesn't provide user-customizations of a cluster objects that is not supported by the operator which owns the object.
+
+## Overview
+
+Render has 3 steps:
+
+1. Validate InstallConfig object. If `install-config.yaml` file is not present fallback to init phase to generate InstallConfig object.
+
+2. Generate all the required TLS assets for the cluster.
+
+3. Use `operator render self` interface to generate the cluster objects for all the operators.
+
+Render puts all the assets on disk for user customization or next phase of the installer.
+
+## Detailed design
+
+### Input
+
+Render phase accepts a valid `InstallConfig` object that exists at `install-config.yaml` in the current directory or the directory specified by `--assets-dir` flag.
+
+### InstallConfig Validation
+
+Render needs to perform 3 kinds of validation:
+
+1. Validate configuration options are known. e.g. networking provider is known.
+
+2. Platform configuration validation. This also includes validating all node pools belong to the specified platform.
+
+3. Ensure `master` node pool exists and has valid configuration.
+
+### TLS asset generation
+
+Generate all the TLS assets required for the cluster and thier corresponding Kubernetes secrets. All this needs to be go code rather than terraform templates.
+
+Some of the TLS assets that need to be generated are:
+1. root-ca
+2. kube-ca
+3. aggregator-ca
+4. etcd-ca
+5. apiserver
+6. ingress and more
+
+### Rendering operator manifests
+
+1. Historically, the installer had manifests for all the cluster objects that are required to run the operator. For example, deployment manifest for the operator, service account, roles/rolebindings and operator config etc. So changes to the installer were required to test any change in the operator. This meant that the first indication if the operator had broken install was after the operator changes were merged in its repository and then operator was updated in the installer repository.
+
+2. Also, all the cluster objects required by the operator to functions had to be managed by the installer team.
+
+Therefore, to allow operator teams to iterate and control all aspects of an operator, the installer needs a process to delegate the `InstallConfig` object to each operator in render phase to generate iteself and all other cluster objects necessary for it to function correctly.
+
+#### Prerequesites
+
+1. Render cannot assume contanier runtime on host.
+2. Render needs to be successfull on atleast linux and darwin systems.
```

Comment:
Maybe "on atleast linux and darwin systems" -> "on (at least) Linux and Darwin systems"?